src/macaron/artifact/local_artifact.py

```diff
@@ -0,0 +1,275 @@
+            root_dir=local_maven_repo,
+            pathname=pattern,
+        )
+
+        for found_path in found_paths:
+            full_path = os.path.join(local_maven_repo, found_path)
+            if os.path.isdir(full_path):
+                artifact_paths.append(full_path)
+
+    return artifact_paths
+
+
+def find_artifact_paths_from_python_venv(
+    venv_path: str,
+    glob_patterns: list[str],
+) -> list[str] | None:
+    """Find python artifacts within a python virtual environment directory.
+
+    For packages in the virtual environment, we will treat their name case-insensitively.
+    https://packaging.python.org/en/latest/specifications/name-normalization/
+
+    Parameters
+    ----------
+    local_maven_repo: str
+        The path to the directories to find artifacts.
+    glob_patterns: list[str]
+        The list of glob patterns that matches to artifact file names.
+
+    Returns
+    -------
+    list[str] | None
+        The list of path to found artifacts in the form of ``local_maven_repo``/<artifact_specific_path>
+    """
+    if not os.path.isdir(venv_path):
+        return None
+
+    artifact_paths = []
+
+    try:
+        venv_path_entries = os.listdir(venv_path)
+    except (NotADirectoryError, PermissionError, FileNotFoundError):
+        return None
+
+    all_package_dirs: list[str] = []
+    for entry in venv_path_entries:
+        entry_path = os.path.join(venv_path, entry)
+        if os.path.isdir(entry_path):
+            all_package_dirs.append(entry)
+
+    for package_dir in all_package_dirs:
+        for pattern in glob_patterns:
+            if fnmatch.fnmatch(package_dir.lower(), pattern.lower()):
+                full_path = os.path.join(venv_path, package_dir)
+                artifact_paths.append(full_path)
+
+    return artifact_paths
+
+
+def _get_local_artifact_path_for_build_tool_purl_type(
+    purl: PackageURL,
+    build_tool_purl_type: str,
+    local_artifact_repo: str,
+) -> list[str] | None:
+    """Find local artifacts within ``local_artifact_repo`` depending on the purl type."""
+    if build_tool_purl_type == "maven":
+        maven_artifact_patterns = construct_local_artifact_paths_glob_pattern_maven_purl(purl)
+        if not maven_artifact_patterns:
+            return None
+
+        artifact_paths = find_artifact_paths_from_local_maven_repo(
+            local_maven_repo=local_artifact_repo,
+            glob_patterns=maven_artifact_patterns,
+        )
+
+        if artifact_paths:
+            return artifact_paths
+
+    if build_tool_purl_type == "pypi":
+        pypi_artifact_patterns = construct_local_artifact_paths_glob_pattern_pypi_purl(purl)
+        if not pypi_artifact_patterns:
+            return None
+
+        artifact_paths = find_artifact_paths_from_python_venv(
+            venv_path=local_artifact_repo,
+            glob_patterns=pypi_artifact_patterns,
+        )
+
+        if artifact_paths:
+            return artifact_paths
+
+    return None
```

Comment:
```suggestion
    if build_tool_purl_type == "maven":
        maven_artifact_patterns = construct_local_artifact_paths_glob_pattern_maven_purl(purl)
        if not maven_artifact_patterns:
            return None

        return find_artifact_paths_from_local_maven_repo(
            local_maven_repo=local_artifact_repo,
            glob_patterns=maven_artifact_patterns,
        )
    elif build_tool_purl_type == "pypi":
        pypi_artifact_patterns = construct_local_artifact_paths_glob_pattern_pypi_purl(purl)
        if not pypi_artifact_patterns:
            return None

        return find_artifact_paths_from_python_venv(
            venv_path=local_artifact_repo,
            glob_patterns=pypi_artifact_patterns,
        )

    return None
```
A log message of some kind might also be nice for unsupported purl types.

Reply:
Good idea. I will add a log message for unsupported purl types.

Regarding the suggested change, this is also affected by https://github.com/oracle/macaron/pull/864#discussion_r1875343210. 

Overall, i think it's a good idea to do the return similar as 
```python
return find_artifact_paths_from_python_venv(
    venv_path=local_artifact_repo,
    glob_patterns=pypi_artifact_patterns,
)
```
So that we make sure that if errors happen in `find_artifact_paths_from_*`, this function will return None.

Reply:
I can agree with that. And in that case, the `elif` isn't really needed in my suggestion.

Reply:
Addressed in https://github.com/oracle/macaron/pull/864/commits/7adbee2cbd4ee972af55aefc50f7acd4fe6214c2